### ISSUE TITLE
Privacy Mode: Failed to get hwnd after started

### ISSUE THREAD

**Issue by TOMIPortugal** [rustdesk/rustdesk#9720]
### Bug Description

On some of our devices running Windows 10, we cannot enable Privacy Mode 1 (the gray screen one), as it fails with one of the following errors:
- "No privacy window created"
- "Failed to get hwnd after started"

Unclear if there are any error logs that we can provide.
Also, we're using RustDesk Server Pro.

### How to Reproduce

1. Connect to a device running Windows 10
1. Enable Privacy Mode 1

### Expected Behavior

Privacy Mode is enabled successfully and hides the sensitive tasks we have to perform.

### Operating system(s) on local side and remote side

Windows 11 -> Windows 10

### RustDesk Version(s) on local side and remote side

1.3.1 -> 1.3.1 (Custom Client)

### Screenshots

![rustdesk_cA1pmYxLNF](https://github.com/user-attachments/assets/c9e65827-c5c6-41c4-82c5-eec1b720f54f)

### Additional Context

We can consistently reproduce this issue on some machines running Windows 10, but not all.

Is there any hardware / windows version requirements for using Privacy Mode?

**Comment by rustdesk** [comment]
Please noted, even this is fixed, privacy mode can only work on one monitor, it does not work for mulitple monitors case. https://github.com/rustdesk/rustdesk/issues/6707#issuecomment-1860073303

**Comment by TOMIPortugal** [comment]
> Please noted, even this is fixed, privacy mode can only work on one monitor, it does not work for mulitple monitors case. [#6707 (comment)](https://github.com/rustdesk/rustdesk/issues/6707#issuecomment-1860073303)

We were already aware of this limitation. The machines we tried this on only had one monitor.

-R

**Comment by fufesou** [comment]
@TOMIPortugal Hi, could you please

1. Open the command line "Win" + "R" -> input "cmd" -> Enter.
2. Run `tasklist | findstr RuntimeBroker_rustdesk.exe`, to check if there're processes for the privacy window?
3. Check if there're any related logs in Event viewer. "Win" -> input "Event Viewer"

![image](https://github.com/user-attachments/assets/e7d09abc-880f-4567-9e20-f008e4f4e216)

4. Compare the installation paths of RustDesk to see if there are any differences, such as the English-only path or the path that includes other languages.

**Comment by fufesou** [comment]
> Run `tasklist | findstr RuntimeBroker_rustdesk.exe`, to check if there're processes for the privacy window?

Could you please run `tasklist | findstr RuntimeBroker_` after trying the privacy mode? 

`tasklist | findstr RuntimeBroker_rustdesk.exe` may not be good, because it does not print `RuntimeBroker_rustdesk.ex` when the process name is a little long.

Or can you please try find the process in the TaskManager?

> Only this, which seems unrelated at first glance.

Seems to be not releated.

> On the failing machines, the first time we try to enable Privacy Mode, the error that shows up is always "Failed to get hwnd after started", which takes about 3 seconds to show up.

Yes. The process will wait a longer and give a different message on first time.


> We also tried to see if the issue was related to specific Windows 10 versions, but the result was inconclusive:

I have no clue for now. Are the failing machines slower or busier? Not sure if the default wait timeout is enough.

**Comment by TOMIPortugal** [comment]
Good catch on the process name being too long! Yes then, there is a process:
![rustdesk_VADdK7zM2C](https://github.com/user-attachments/assets/2e5e2985-b160-4097-b3bd-1a79ff4bcd20)

Also, all the machines I tested on the previous message had the exact same CPU and RAM specs. They're about equally as busy.

-R

**Comment by fufesou** [comment]
> there is a process:

Maybe there're some unknown reasons (maybe permissions) for creating a new font most window.

Could you please also try https://github.com/fufesou/RustDeskTempTopMostWindow/releases/tag/v0.3 on your failing  machines?

It will create a front most window to cover the display, then exit after about 5 seconds if it runs normally.

**Comment by TOMIPortugal** [comment]
> Could you please also try https://github.com/fufesou/RustDeskTempTopMostWindow/releases/tag/v0.3 on your failing machines?

I tried running this on one of the working machines on our office and, unless we're doing something wrong, the TestApp.exe application does not work:

```
➜ TempTopMostWindow_x86  ./TestApp.exe
Failed CreateProcess, 0x2, O sistema não conseguiu localizar o ficheiro especificado.
Failed FindWindow, 0x2, O sistema não conseguiu localizar o ficheiro especificado.
```
Translation: The system was unable to locate the specified file.

The TestAppDebug.exe application does work, but it does not exit by itself at all.

> Please run with a physical monitor. The remote control application can see nothing in this scenario, it will not see the top most window.

Unfortunately this isn't so simple on our end given the devices are far away from. All of the machines we have on our office are working.

-R

**Comment by fufesou** [comment]
> TestApp.exe application does not work

Do you run the x86 executable on x64 machine?

> The TestAppDebug.exe application does work, but it does not exit by itself at all.

Same behavior for me. This needs to be changed, it should exit after 5 seconds.

> All of the machines we have on our office are working.

Do you mean the test app and RustDesk privacy mode work fine on thoes machines?

> Unfortunately this isn't so simple on our end given the devices are far away.

I've tested `TestAppDebug.exe` it can create a top most window, and I can see it from a remote control application.

Maybe you can try it for simple test, though it does not exit after 5 seconds.

You can
1. Hover the cursor on the taskbar.
1. Find the newly created window
2. Close it

**Comment by TOMIPortugal** [comment]
> Do you run the x86 executable on x64 machine?

Yes. Didn't think it would make a difference. When running the x64 version of the TestApp, the Privacy Mode didn't show up but the error message was different:
```
➜ TempTopMostWindow_x64  ./TestApp.exe
Failed FindWindow, 0x0, A operação foi concluída com êxito.
```
Translation: The operation completed successfully 😕
 
> Do you mean the test app and RustDesk privacy mode work fine on thoes machines?

Yes. TestApp.exe is giving errors but the RustDesk Privacy Mode works correctly on these machines.
TestAppDebug.exe works on those machines though.

> Maybe you can try it for simple test, though it does not exit after 5 seconds.

The TestAppDebug.exe application doesn't seem to work on the Devices where the Privacy Mode also doesn't work. On the contrary, on the Devices where the Privacy Mode works, the TestAppDebug also works.

On the Devices where it doesn't work, neither applications show any logs:
![rustdesk_uhe53ddozq](https://github.com/user-attachments/assets/f52134b5-c739-4720-9d68-97075c5365f4)

*** 

So, to wrap up:

On machines where the Privacy Mode **does** work:
- TestApp.exe: does not work, and exits with "Failed FindWindow, 0x0, The operation completed successfully."
- TestAppDebug.exe: works

On machines where the Privacy Mode **does not** work:
- TestApp.exe: exits without logs
- TestAppDebug.exe: exits without logs

**Comment by fufesou** [comment]
> ```
> ➜ TempTopMostWindow_x64  ./TestApp.exe
> Failed FindWindow, 0x0, A operação foi concluída com êxito.
> ```

I've read the code, the logic is correct. As the message says, the topmost window cannot be found, but no error is detected.

> TestApp.exe is giving errors but the RustDesk Privacy Mode works correctly on these machines. TestAppDebug.exe works on those machines though.

The most possible reason is related to the security settings or permission settings.

| app |  signed | dll injection |
| :--: | :--: | :--: |
| RustDesk | Y | Y |
| TestApp.exe | N | Y |
| TestAppDebug.exe | N | N |

But I'm still trying to find the settings.

**Comment by fufesou** [comment]
@TOMIPortugal Hi, could you please run this test on the failed machine again https://github.com/fufesou/RustDeskTempTopMostWindow/releases/tag/test_find_window

Please only run the `TestApp.exe`.

It changes nothing but write logs in `a.txt` in current directory.

https://github.com/rustdesk-org/RustDeskTempTopMostWindow/compare/main...fufesou:RustDeskTempTopMostWindow:test/debug_logs?expand=1

**Comment by TOMIPortugal** [comment]
I tried running it on a device where Privacy Mode works and another where it doesn't, and the output was the same.

```
UwU: UwU() is called
UwU: BitmapLoader is initialized
```

-R

**Comment by fufesou** [comment]
Please try https://github.com/fufesou/RustDeskTempTopMostWindow/releases/tag/test_find_window again. I've added more logs.

**Comment by TOMIPortugal** [comment]
On the machine where the Privacy Mode doesn't work, the logs were the same:
```
UwU: UwU() is called
UwU: BitmapLoader is initialized
```

On my machine, where the Privacy Mode works, these were the logs:
```
UwU: UwU() is called
UwU: BitmapLoader is initialized
CreateWin: RegisterClassEx is called
CreateWin: LoadLibrary is called
CreateWin: GetProcAddress is called
```

-R

**Comment by fufesou** [comment]
Could you please run some more times? Because the logs seem not to break at the same place.

**Comment by fufesou** [comment]
Could you please also check if "Privacy mode" of this build works? https://github.com/fufesou/rustdesk/releases/download/nightly/rustdesk-1.3.2-x86_64.exe

This is the same as the official one, but without the signature.

**Comment by fufesou** [comment]
Please try https://github.com/fufesou/RustDeskTempTopMostWindow/releases/tag/test_find_window again.

And run `TestAppDebug.exe` on the machine where "Privacy Mode" does not work. It will exit after 3 seconds and write some logs now.

**Comment by TOMIPortugal** [comment]
For the nightly build, this appears as soon as I run the executable:
![image](https://github.com/user-attachments/assets/7af1468a-b8fc-40c8-91f9-ccd388bbe79d)


For TestAppDebug.exe, it did work this time! The Privacy Mode window showed up while connected to the device via rustdesk. Here are the logs either way:

```
UwU: UwU() is called
UwU: BitmapLoader is initialized
CreateWin: RegisterClassEx is to be called
CreateWin: RegisterClassEx is called: rse is 49754
CreateWin: LoadLibrary is to be called
CreateWin: LoadLibrary is called
CreateWin: GetProcAddress is to be called
CreateWin: GetProcAddress is called: pCreateWindowInBand address is 00007FFCCC6E58B0
CreateWin: pCreateWindowInBand is called: hwnd address is 00000000000105E2
CreateWin: SetWindowText is to be called
CreateWin: SetWindowText is called
CreateWin: MonitorFromWindow is called: hmon is 0000000000010001
CreateWin: GetMonitorInfo is to be called
CreateWin: GetMonitorInfo is called
CreateWin: SetWindowPos is to be called
CreateWin: SetWindowPos is called
CreateWin: SetWindowLong is to be called
CreateWin: SetWindowLong is called
CreateWin: ShowWindow is called
CreateWin: UpdateWindow is called
UwU: Window is created
UwU: GetClientRect is called
UwU: CreateDIBFromFile is called
UwU: ShowWindow is called
```

**Comment by fufesou** [comment]
> A referral was returned from the server

The most possible reason is because the executable is not signed. Maybe you can just delete it. It's not worth it to modify your computer's security settings just to try this.

I get this error today. But the following link does not solve it.

Just for ref

https://answers.microsoft.com/en-us/windows/forum/all/always-getting-a-referral-was-returned-from-the/8a5802c8-0815-4abc-aab2-ed49dbbfb971


> For TestAppDebug.exe, it did work this time!

So it does not work one day ago, but now it works?
I've read the code and it should always work. It just show window with normal order. Not sure why it didn't work before.

The most possible reason is still the secure policy that prevent dll inject on some computers.

Sorry, I still can't find the setting.
I have to prioritize other work, so I may not find a solution to this problem in the short term, so I have to put it aside for now.

**Comment by TOMIPortugal** [comment]
> So it does not work one day ago, but now it works?

Yep. Seems to be working consistently everytime after your latest changes.

**Comment by fufesou** [comment]
> Seems to be working consistently everytime after your latest changes.

But I didn't make any changes except some logging. 😅

**Comment by twprh** [comment]
maybe a antivirus software kills some things ???
just a thought !???

**Comment by TOMIPortugal** [comment]
We completely reset one of the PCs where the Privacy Mode 1 was not working, to ensure its not one of our customizations that stops it from working.

Brand new Windows 11, Privacy Mode 1 still did not work. Got the same errors as before.

**Comment by rustdesk** [comment]
We have spended several days on this, but failed to reach a good solution. Sorry, we have to lower its priority.
